### PR TITLE
[FEAT] Support ReturningSideEffects and dispatch<AnyStateUpdater>

### PR DESCRIPTION
**Pull request by cipolleschi**
This pr adds the support for a partially erased `SideEffect` where the only thing typed is the `ReturnValue`.
It also add the support to dispatch `AnyStateUpdater` instead of only the `StateUpdater`. 

This changes allows libraries (and Ramen specifically) to keep the `Dispatchable` protocols internal without the need to expose them to the apps.

Tests builds and pass correctly.

### CODE REVIEW

File: Katana/Store.swift [hunk 1]
```diff
@@ -339,6 +359,34 @@ open class Store<S: State, D: SideEffectDependencyContainer>: PartialStore<S> {
...
+
+   The `Store` follows strict rules about the parallelism with which dispatched items are handled.
+   At the sime time, it tries to leverages as much as possible the modern multi-core systems that our
+   devices offer.
+
+   When a `ReturningSideEffect` is dispatched, Katana will handle them in a parallel queue. A `ReturningSideEffect` is executed and considered
+   done when its body finishes to be executed. This means that side effects are not guaranteed to be run in isolation, and you
+   should take into account the fact that multiple side effects can run at the same time. This decision has been taken to greately
+   improve the performances of the system. Overall, this should not be a problem as you cannot really change
+   the state of the system (that is, the store's state) without dispatching a `ReturningStateUpdater`.
```
Comment (albertogiunta) [inline Katana/Store.swift]: `ReturningStateUpdater` :notme:

File: Katana/Store.swift [hunk 2]
```diff
@@ -339,6 +359,34 @@ open class Store<S: State, D: SideEffectDependencyContainer>: PartialStore<S> {
...
+
+   #### Promise Resolution
+
+   When it comes to `ReturningSideEffect`s, the promise is resolved when the body of the `RetunringSideEffect` is executed entirely (see
```
Comment (albertogiunta) [inline Katana/Store.swift:380]:
```suggestion
   When it comes to `ReturningSideEffect`s, the promise is resolved when the body of the `ReturningSideEffect` is executed entirely (see
```

File: Katana/SideEffect.swift [hunk 3]
```diff
@@ -44,15 +44,26 @@ public protocol AnySideEffectContext {
...
+  - parameter dispatchable: the side effect to dispatch
+  - returns: a promise that is resolved with the value returned by the side effect
+             when the store finishes handling the dispatched item
+  */
+  @discardableResult
+  func dispatch<T: ReturningSideEffect>(_ dispatchable: T) -> Promise<T.ReturnValue>
   
   /**
-   Dispatches a `StateUpdater`. It the type-safe version of the `anyDispatch` method
+   Dispatches a `AnyStateUpdater`. By considering that all the `StateUpdater`s are also `AnyStateUpdater`s this dispatch
```
Comment (AlfioRusso) [inline Katana/SideEffect.swift:59]:
```suggestion
   Dispatches an `AnyStateUpdater`. By considering that all the `StateUpdater`s are also `AnyStateUpdater`s this dispatch
```

File: Katana/Store.swift [hunk 4]
```diff
@@ -349,20 +397,20 @@ open class Store<S: State, D: SideEffectDependencyContainer>: PartialStore<S> {
    At the sime time, it tries to leverages as much as possible the modern multi-core systems that our
    devices offer.
    
-   When a `StateUpdater` is dispatched, the Store enqueues it in a serial and syncronous queue. This means that the Store
+   When an `AnyStateUpdater` is dispatched, the Store enqueues it in a serial and syncronous queue. This means that the Store
    executes one update of the state at the time, following the order in which it has received them. This is done
    to guarantee the predictability of the changes to the state and avoid any race condition. In general, using a syncronous
-   queue is never a big problem as any operation that goes in a `StateUpdater` is very lighweight.
+   queue is never a big problem as any operation that goes in an `AnyStateUpdater` is very lighweight.
```
Comment (AlfioRusso) [inline Katana/Store.swift:403]:
```suggestion
   queue is never a big problem as any operation that goes in an `AnyStateUpdater` is very lightweight.
```

File: Katana/Store.swift [hunk 5]
```diff
@@ -339,6 +359,34 @@ open class Store<S: State, D: SideEffectDependencyContainer>: PartialStore<S> {
   override public func dispatch<T: SideEffect>(_ dispatchable: T) -> Promise<T.ReturnValue> {
     return self.enqueueSideEffect(dispatchable)
   }
```
Comment (AlfioRusso) [inline Katana/Store.swift:361]: do we still need this? or can we drop it? same for context function.

Comment (cipolleschi) [reply Katana/Store.swift:361]: I tested without the `SideEffect` in `Ramen` and `RamenDemo` and it works! :D I'm removing them in the next commit to keep the API as simple as possible! ;)

**Pull request merged**